Use polyfill for Object.assign (rather than alternative name)

sbc100 opened the pull request:
This gives a small code size saving and it makes our codebase
more readable.

The reason for this change is that I ran into an issue where I was
trying to use `objAssign` in more places and it became apparent
that MINIMAL_RUNTIME didn't define the `objAssign` helper.

Also, move the defintion of the polyfill inside of `#if POLYFILL` so
it can be disabled using `-sNO_POLYFILL`.

See #15823 #15938

juj reviewed (approved):
Great!

(if Object.assign() is referenced a lot, then using a custom name `objAssign()` will win in code size, since Closure is not smart enough to outline/minify function accesses.. though looks like it isn't referenced often enough here)

kripken commented on src/polyfill/objassign.js:
```diff
@@ -0,0 +1,28 @@
+ * @return {!Object} The target object is returned.
+ */
+if (Object.assign === 'undefined') {
+  Object.assign = function(target, source) {
```
(When there is just one use then `var objAssign = Object.assign` does add size.)

sbc100 replied:
Regarding the code size issue,  I agree that we should consider adding our own short names if/when it makes sense, but we should not single out `Object.assign` for this treatment.  For example we could say exacatly the same thing about other global names such as `WebAssembly.instantiateStreaming`.

Regarding polyfilling, I agree we should come up with policy.   At the very least the user should be able to disable polyfils if they wish.   I will open an issue discussion.

curiousdannii replied:
Polyfills can be added by other build tools. IMO if Emscripten adds any itself it should be opt-in, not opt-out.

sbc100 replied:
I'm tempted to go with opt-in so things just work by default... but I could be persuaded otherwise.    We have historically tended to assume that our output should "just work".   We can followup on #15938.

sbc100 replied:
We also have a historical precedent here that we have been poly-filling `Promise` on older platforms for while now and nobody has asked us about an opt out (possibly because very few users are targeting these older platforms).   Another option: Force users to opt in or out explicitly when older browser is targeted.

curiousdannii replied:
Are polyfills only added if you manually set old browsers as supported? That basically is opting in.

sbc100 replied:
Exactly, yes.  By default I don't think we would want any polyfills in the output (and we don't have any today).

sbc100 commented:
I updated this PR to honor the new NO_POLYFILL setting (#16036), and also rebased all the codesize improvements.

If we want to discuss the possibility of using short names for things like `Object.assign` I think that we can continue to do that in #15938, but we can land this for now since we get codesize win.

kripken commented on src/shell.js:
```diff
@@ -43,20 +43,11 @@ var Module = typeof {{{ EXPORT_NAME }}} !== 'undefined' ? {{{ EXPORT_NAME }}} :
-}
-#else
-var objAssign = Object.assign;
+#include "polyfill/objassign.js"
```
Should this not be guarded behind `#if POLYFILL` so that flag is respected?

sbc100 replied:
It is yes, the #endif for the POLYFILL ifdef was moved from line 46 down to line 50.  

I'll add a comment to make it clear

kripken replied:
Ah, thanks, now I see it...
